fix: Cron Job to ping code-owners who has not reviewed assigned PRs within 24 hrs.

Naman-1234 opened the pull request:
<!--
  - Thanks for submitting code to Oppiabot! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Basically, this Pr is related to assign code owner in case that PR is not being reviewed under 24 hrs of Code Push.
Fixes #182 
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

Naman-1234 commented:
@jameesjohn, Sir  Could you please review and let me know if I am going right till now and Could you tell me how can I test whether I am going in a right way or not(As Usually in another nodejs or javascript based project, I take help of console.log to know the outcome till that point).
Further,I have thought of approaching this problem as:
1. First Find all Open Pull Requests(Already Done).
2. Filter out having those who do not have any changes requested.
3. Then going to the above Pr(Which We get in the 2nd point without any changes requested) and check its time period of last updated by the assignee and check if it is more than 24 hrs.
4. After that going through all Pr satisfying 3rd and mention code owners(Getting From Labels ) to review that PR.

Please let me know if there is any problem with above approach.Thanks.

Naman-1234 commented:
@vojtechjelinek  Sir Could you please review this PR.

jameesjohn commented:
Hi @Naman-1234 

Your stated approach looks nice, here are my thoughts

> Filter out having those who do not have any changes requested.

I don't think we need to filter these out. If the code-owners have been assigned to the PR already, then their review is required.

> After that going through all Pr satisfying 3rd and mention code owners(Getting From Labels ) to review that PR.

Since the code owners have already been assigned to the PR, you can get them from the list of assignees of the PR.

Naman-1234 commented:
@jameesjohn Ok Sir , Surely I will try that approach, And Sir Could you tell me how can I test , Means suppose I am saying:
const x=someFunction() then how can i know the value of x in this case. As in normal javascript we take help of console.log

jameesjohn commented:
Yes, you can test on other repositories as well.
Console log statements work here as well, as it is just javascript.
You might need to test your code locally by setting the periodic check to run in a few minutes so that you don't wait for 24 hours.
The periodic check event will get triggered immediately the bot is run, and then at every other interval specified by you. https://github.com/oppia/oppiabot/blob/72eacc99b68b82d2d14f73b1362eaaff3536ec8a/index.js#L195.

Naman-1234 commented:
@jameesjohn  Sir Sorry I am disturbing you again but I just have a little query. How should I connect my js file written to index.js to cover it in this periodic checks and also that periodic check will cover for all files but For testing I only want it for My file. Could you help me out with this.

jameesjohn commented:
You can export your function, then import it in the index.js file, and add it to this array:
https://github.com/oppia/oppiabot/blob/72eacc99b68b82d2d14f73b1362eaaff3536ec8a/index.js#L145-L149

Naman-1234 commented:
@jameesjohn Sir Sorry for interruption, I did as told. Exported it and added it to our array. Also added alert statements. But they didn't show up. Could you tell me where I have gone wrong?

Naman-1234 commented:
@jameesjohn  and @vojtechjelinek  Sir I have written a function. And tested it if it works as you can see in the below Screenshot. Could you guide me how can i write Test file (Those same as spec folder). Currently I tested in my own repo through use of another github Id in order to create pull requests etc. Also could you tell me why one of check is failing. So that I can fix that.
Here is the screenshot. Please review this.Thanks.


![Screenshot (1203)](https://user-images.githubusercontent.com/63748249/112376642-b6250180-8d0a-11eb-9126-493a1dc7ff27.png)

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  Sir Could you please review this PR. I have also added the test . Please let me know if there are any changes requested.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  PTAL Sir. I think this issue is been targeted. Please let me know if there are any changes requested. I have also written the tests and also manually tested it, Please find screenshots of them below. Thanks.

### Here Sahaayak is one of my repo in which I used my another Github Id for testing.
<div align="center">
<img src="https://user-images.githubusercontent.com/63748249/112521630-bc76b480-8dc2-11eb-96fe-69113660e8d5.png"></img><br></br><br></br><br></br>
<img src="https://user-images.githubusercontent.com/63748249/112521619-ba145a80-8dc2-11eb-91b1-62c0f2f7a3c1.png"></img>
</div>

jameesjohn commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,76 @@
+ * @fileoverview Handler for checks to be carried out periodically.
+ */
+
+const { context } = require('@actions/github');
```
This is not required.

Naman-1234 replied:
Ok Sir Will remove that.

jameesjohn commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,76 @@
+  timeOfPr = new Date(timeOfPr);
+  let currentTime = new Date.now();
+  let diff = Math.abs(timeOfPr - currentTime);
+  diff > 24 ? true : false;
```
Store `24` in a constant variable

Naman-1234 replied:
Will do that Sir.Thanks,.

jameesjohn commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,76 @@
+
+
+//This is a function to check if there are more than 24 Hours with the Pr
+const checkTime = (pullRequest)=>{
```
This function should probably return a boolean.
Currently, it returns undefined.

Naman-1234 replied:
Ok Sir. WiIl replace by the statement you tell in below suggestion.

jameesjohn commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,76 @@
+  timeOfPr = new Date(timeOfPr);
+  let currentTime = new Date.now();
+  let diff = Math.abs(timeOfPr - currentTime);
+  diff > 24 ? true : false;
```
```suggestion
  return diff > 24;
```

Naman-1234 replied:
Ok Sir.

jameesjohn commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
Can this functionality be moved to the periodic check module?

Naman-1234 replied:
Sir, It can be moved, I can do that if you want. Initially, I also thought of doing the same but then I think its a separate functionality just like all the others in the lib folder, That's why I made a different File for that. I would like to know your suggestions on this one sir. If you say then I can transfer the same to periodic check module.

jameesjohn commented on index.js:
```diff
@@ -192,6 +197,7 @@ const checkAuthor = (context) => {
 module.exports = (oppiabot) => {
   scheduler.createScheduler(oppiabot, {
     delay: !process.env.DISABLE_DELAY, // delay is enabled on first run
+    //* Changing Temporarily to 10000sec in order to get better testing
```
This should be removed.

Naman-1234 replied:
Will remove that Sir.

jameesjohn reviewed (changes requested):
Thanks, @Naman-1234, I left some comments.

Naman-1234 commented:
@jameesjohn  and @vojtechjelinek  Could you tell me the reason behind failing of above test. I would like to solve that and also could you tell me why is pullRequest coming undefined in my local repository(Bcoz it is being used at so many places and they all kind of giving errors for that). Please find below screenshots and let me know the reason please.

### Here the code which you are seeing is from original repo, and nothing is changed from my side except that console statements, but still it is saying pullrequest to be undefined.
<br></br>
![Screenshot (1208)](https://user-images.githubusercontent.com/63748249/112610515-cb04b080-8e42-11eb-81c9-5a3f4af667fd.png)




![Screenshot (1209)](https://user-images.githubusercontent.com/63748249/112610526-cfc96480-8e42-11eb-88f6-b01a18f1fca7.png)

Naman-1234 commented:
@vojtechjelinek  PTAL Sir.Thanks.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
Rename this file to exclude the 24 hours, so something like pingCodeOwners.js

vojtechjelinek replied:
```suggestion
// Copyright 2021 The Oppia Authors. All Rights Reserved.
```

Naman-1234 replied:
On it Sir.

vojtechjelinek commented on index.js:
```diff
@@ -44,6 +44,10 @@ const whitelistedAccounts = (process.env.WHITELISTED_ACCOUNTS || '')
   .toLowerCase()
   .split(',');
 
+const checkAfter24Hrs = require(
+  './lib/pingCodeOwnerAfter24hrs'
+);
```
Move this more up with the other imports.

Naman-1234 replied:
Ok Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+const ADay = 24;
```
Rename it to something more appropriate, from the name it should be understood that this is the period after which review is required.

Naman-1234 replied:
Naming it to reviewPeriod

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+
+
+const checkAnyReviewRequiredPr = async(context)=>{
+  await sleep(5000);
```
Why the `sleep` here? Please add a comment explanation.

Naman-1234 replied:
It was for testing purpose. Will remove that Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+
+const checkAnyReviewRequiredPr = async(context)=>{
+  await sleep(5000);
+  console.log(await getAllOpenPullRequests(context));
```
Remove this if you don't need it for anything.

Naman-1234 replied:
Sure Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+  await sleep(5000);
+  console.log(await getAllOpenPullRequests(context));
+  const allOpenPullRequests = await getAllOpenPullRequests(context);
+  const temp = context.github.pulls;
```
What is `temp` use better name.

Naman-1234 replied:
Removed it, because it was used for console purpose eariler. Forgot to remove that.Sorry for that Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+
+module.exports = {
+  checkAnyReviewRequiredPr
+};
```
Add newline at the end of the file.

Naman-1234 replied:
Sure Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+  const allOpenPullRequests = await getAllOpenPullRequests(context);
+  const temp = context.github.pulls;
+  const requiredPrs = await allOpenPullRequests.filter(
+    (pullRequest)=>{
```
```suggestion
    (pullRequest) => {
```

Naman-1234 replied:
Sure Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+  return diff > Aday;
+};
+
+const pingCodeOwners = async(context, pullRequest)=>{
+  const reviewers = pullRequest.requested_reviewers;
+  const message = 'Hi' + reviewers.map((reviewer)=>{
```
```suggestion
const pingCodeOwners = async(context, pullRequest) => {
  const reviewers = pullRequest.requested_reviewers;
  const message = 'Hi' + reviewers.map((reviewer) => {
```

Naman-1234 replied:
Done Sir.

vojtechjelinek commented on lib/pingCodeOwnerAfter24hrs.js:
```diff
@@ -0,0 +1,71 @@
+const pingCodeOwners = async(context, pullRequest)=>{
+  const reviewers = pullRequest.requested_reviewers;
+  const message = 'Hi' + reviewers.map((reviewer)=>{
+    return ' @' + reviewer.login + ' ';
+  }) + ', Its been more than 24 Hrs, ' +
+   'since Requested Review. Please Review this. Thanks';
```
```suggestion
    return ' @' + reviewer.login + ' ';
  }) + ', Its been more than 24 hours, ' +
   'since the review was requested. Please review this. Thanks!';
```

Naman-1234 replied:
Done Sir.

vojtechjelinek commented on spec/pingCodeOwnerAfter24hrsSpec.js:
```diff
@@ -0,0 +1,102 @@
+      expect(github.issues.createComment).not.toHaveBeenCalled();
+    });
+  });
+});
```
Missing empty line at the end of the file.

Naman-1234 replied:
Adding it Sir.Thanks.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few initial comments.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  Sir. Changes Requested have been targeted. Please review this.Thanks.

vojtechjelinek commented:
Please do not mark comments as resolved by yourself, just reply to them and let the reviewer handle them.

Naman-1234 commented:
> Please do not mark comments as resolved by yourself, just reply to them and let the reviewer handle them.

I am so sorry for that. I didn't know about that. This will not be done  in future. Again extremely sorry Sir.

jameesjohn commented:
> Sir, It can be moved, I can do that if you want. Initially, I also thought of doing the same but then I think its a separate functionality just like all the others in the lib folder, That's why I made a different File for that. I would like to know your suggestions on this one sir. If you say then I can transfer the same to periodic check module.


On second thoughts, it should be moved to the pull request review module since it has to do with checking if the code owners have reviewed the pull requests.

What you could do is to create a new function that would contain the functionality in the pull request review module, export it, and add it to be called on the periodic check event.

jameesjohn reviewed (changes requested):
Thanks, @Naman-1234, I took a second pass. PTAL.

Naman-1234 commented on lib/pingCodeOwner.js:
```diff
@@ -0,0 +1,69 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+const reviewPeriod = 24;
```
Thanks for mentioning that Sir.I forgot to change its name where it is being used, Sorry for that.

Naman-1234 commented on lib/pingCodeOwner.js:
```diff
@@ -0,0 +1,69 @@
+  timeOfPr = new Date(timeOfPr);
+  let currentTime = new Date.now();
+  let diff = Math.abs(timeOfPr - currentTime);
+  return diff > Aday;
```
Will be replaced by reviewPeriod.

Naman-1234 commented on lib/pingCodeOwner.js:
```diff
@@ -0,0 +1,69 @@
+
+//This is a function to check if there are more than 24 Hours with the Pr
+const checkTime = (pullRequest)=>{
+  let timeOfPr = pullRequest.updated_at;
```
Sure Sir.

jameesjohn commented on index.js:
```diff
@@ -44,6 +46,8 @@ const whitelistedAccounts = (process.env.WHITELISTED_ACCOUNTS || '')
   .toLowerCase()
   .split(',');
 
+
+
```
Not required.

Naman-1234 replied:
removing it sir.

jameesjohn commented on index.js:
```diff
@@ -35,7 +35,9 @@ const checkPullRequestReviewModule = require('./lib/checkPullRequestReview');
 const ciCheckModule = require('./lib/ciChecks');
 const periodicCheckModule = require('./lib/periodicChecks');
-
+const checkAfter24Hrs = require(
+  './lib/pingCodeOwner'
+);
```
This functionality should be moved to the pull request review module.

jameesjohn replied:
See https://github.com/oppia/oppiabot/pull/215#issuecomment-808814802

Naman-1234 replied:
Ok sir.

jameesjohn commented on lib/pingCodeOwner.js:
```diff
@@ -0,0 +1,69 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+const reviewPeriod = 24;
```
```suggestion
const MAX_REVIEW_PERIOD = 24;
```

Naman-1234 replied:
Changing it now sir.

jameesjohn commented on lib/pingCodeOwner.js:
```diff
@@ -0,0 +1,69 @@
+  timeOfLastUpdate = new Date(timeOfLastUpdate);
+  let currentTime = new Date.now();
+  let diff = Math.abs(timeOfLastUpdate - currentTime);
+  return diff > reviewPeriod;
```
```suggestion
  return diff > MAX_REVIEW_PERIOD;
```

Naman-1234 replied:
Changing it now sir.

Naman-1234 commented:
@jameesjohn PTAL Sir, As directed by you, It has been transferred to a single function in CheckPullRequestReview File.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn Sir All the Changes requested have been targeted.Could you please review this now. Thanks.

jameesjohn commented on spec/pingCodeOwnerSpec.js:
```diff
@@ -0,0 +1,104 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
```
The test should also be moved to the appropriate test file.

Naman-1234 replied:
On it Sir.

Naman-1234 commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,48 @@ const handleResponseToReview = async (context) => {
+  return diff > MAX_REVIEW_PERIOD;
+};
+
+const checkAnyReviewRequiredPr = async(context) =>{
```
Ok Sir

Naman-1234 commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,48 @@ const handleResponseToReview = async (context) => {
+ */
+
+//This is a function to check if there are more than 24 Hours with the Pr
+const checkTime = (pullRequest) =>{
```
Changing it now Sir.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  PTAL Sir. I think this issue is targeted. Please let me know if any change is required. Thanks.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  PTAL Sir, All the changes requested have been addressed, Please review this. Thanks.

vojtechjelinek commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,49 @@ const handleResponseToReview = async (context) => {
+
+//This is a function to check if there are more than 24 Hours with the Pr
+const isMaxReviewPeriodExceeded = (pullRequest) =>{
+  let timeOfLastUpdate = pullRequest.updated_at;
+  timeOfLastUpdate = new Date(timeOfLastUpdate);
```
Hmm, how do we handle cases when only one reviewer forgets to review but otherwise there is an ongoing activity on the PR.

vojtechjelinek replied:
I think that this is a problem in general, we should ping individual reviewers separately and not all at once since this can cause spam for the other reviewers.

Naman-1234 replied:
Ok sir, in that case, I am writing a similar function but with functionality where if any one of the reviewers have reviewed it till now, Then It will not tag others, Otherwise, it will tag a reviewer(1st one in that reviewers list). Is that correct Sir. please let me know so I can proceed in this Pr. Thanks.

Naman-1234 replied:
@vojtechjelinek Sir Could you please elaborate on this, I am not getting what actually Functionality you are looking for, It will be great if you try to explain or give an example. Thanks.

Naman-1234 replied:
@jameesjohn Sir Have you got what Sir wants to say, If yes could you please provide an explanation, So that I can proceed in this PR. Thanks.

vojtechjelinek replied:
Basically, imagine we have multiple reviewers on a PR, some already approved and now the other reviewers are slowly also approving the PR. 

Firstly, the thing is that we only check `pullRequest.updated_at` meaning that when for example Oppiabot adds a comment it is still registered as an update and the reviewers won't get notified.

Secondly, when there is more than 24 hours without update we then ping which reviewers exactly? The ones that didn't review at all or any other set of reviewers?

Naman-1234 replied:
Ok Sir. Thanks for telling. Now I could proceed with this PR.Thanks.

jameesjohn replied:
To know which reviewers to ping, you can fetch all the reviews for a PR, and determine who is yet to review and to also check when their last reviews were made.

Naman-1234 replied:
Ok Sir, Thanks for guiding me.

vojtechjelinek reviewed (changes requested):
Thanks! Left a comment.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn PTAL Sir, I changed the function to those reviewers only who have not reviewed PR for more than 24 HRS. Please review this and let me know if any changes are requested. Thanks.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn PTAL Sir, As suggested the function is been written and tests written along are also passing. Please review this PR. Thanks.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  Sir,  Functionality has been changed as per the suggestions, and tests are written are also passing. Could you please review this Pr. Thanks.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+  return diff > MAX_REVIEW_PERIOD;
+};
+
+const checkForReviewersWithPendingReview = async(context) => {
```
```suggestion
const checkForReviewersWithPendingReview = async (context) => {
```

Naman-1234 replied:
Will do it in the latest commit, Sir.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+  let diff = Math.abs(timeOfReviewSubmission - currentTime);
+  return diff > MAX_REVIEW_PERIOD;
+};
+
+const checkForReviewersWithPendingReview = async(context) => {
```
Can you add a doc block above this function that describes the function?
See what is done in `handleResponseToReview`

jameesjohn replied:
Also, add the required JS doc comments for each parameter in the function.

Naman-1234 replied:
Sure Sir.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+
+      const allReviewers = [];
+      allReviews.map((review) => {
+        const booleanVariable = isMaxReviewPeriodExceeded(review);
```
Use a more intuitive variable name.

Naman-1234 replied:
Using `isReviewPeriodExceeded` instead.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+        })
+      );
+
+      const allReviewers = [];
```
Maybe rename to `reviewersToPing`. `allReviewers` is somewhat conflicting with `allReviews`.

Naman-1234 replied:
Sure sir.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+        // If Max Period Exceeded then We need to ping that reviewer.
+        if (booleanVariable) {
+          const reviewer = review.user.login;
+          if (!allReviewers.includes(reviewer)) {
+            allReviewers.push(reviewer);
+          }
```
Should you also check if the reviewer is assigned to the pull request?

jameesjohn replied:
Rethinking the issue, I think it will be safer and easier check assigned reviewers and then check when their last review was, if any, and decide to ping them if they fit the 24 hours category.

Naman-1234 replied:
Sir, I think this will not be a problem, As soon code owners will be decided by CODEOWNERS File. So this might become redundant.
Looking forward to your views on this.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+ */
+
+//This is a function to check if there are more than 24 Hours with the Pr
+const isMaxReviewPeriodExceeded = async (review) => {
+  let timeOfReviewSubmission = review.submitted_at;
+  timeOfReviewSubmission = new Date(timeOfReviewSubmission);
+  let currentTime = new Date.now();
+  let diff = Math.abs(timeOfReviewSubmission - currentTime);
+  return diff > MAX_REVIEW_PERIOD;
+};
```
Has this been tested manuallly?

Naman-1234 replied:
Yes sir, I tested it manually and also saw intermediate steps or dates through console statements.

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+const isMaxReviewPeriodExceeded = async (review) => {
+  let timeOfReviewSubmission = review.submitted_at;
+  timeOfReviewSubmission = new Date(timeOfReviewSubmission);
+  let currentTime = new Date.now();
```
I don't think `new Date.now()` works.

Naman-1234 replied:
Sorry for that It should be a new Date() only, as that automatically returns the current DateTime. Thanks for pointing this out

jameesjohn commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,63 @@ const handleResponseToReview = async (context) => {
+  let timeOfReviewSubmission = review.submitted_at;
+  timeOfReviewSubmission = new Date(timeOfReviewSubmission);
+  let currentTime = new Date.now();
+  let diff = Math.abs(timeOfReviewSubmission - currentTime);
+  return diff > MAX_REVIEW_PERIOD;
```
`diff` would be in milliseconds right?
`MAX_REVIEW_PERIOD` is in hours so `diff` will always be greater than `MAX_REVIEW_PERIOD`.

Naman-1234 replied:
This one I missed Sir, I will convert it in milliseconds. Thanks for pointing it out.

jameesjohn reviewed (changes requested):
Thanks, I took a pass.

vojtechjelinek commented:
@Naman-1234 Is this ready for review?

Naman-1234 commented:
@vojtechjelinek  No Sir, One or two small changes are there which I will push either by today or tomorrow. Sorry for the delay. Thanks.

Naman-1234 commented:
@vojtechjelinek  and @jameesjohn  Sir, this PR is ready for review now. Tests are also passing. Thanks.
![Screenshot (166)](https://user-images.githubusercontent.com/63748249/115401079-a0bace80-a207-11eb-9e70-7066cd1dda2b.png)

vojtechjelinek commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,70 @@ const handleResponseToReview = async (context) => {
   }
 };
 
+/**
+ * @param {import('probot').Context} context
+ * @param {string[]} assignees
+ * @param {string} comment
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
+
+//This is a function to check if there are more than 24 Hours with the Pr
```
Please format the docstring properly.

Naman-1234 replied:
Doing that in the latest commit Sir. Thanks.

vojtechjelinek commented on lib/checkPullRequestReview.js:
```diff
@@ -381,9 +382,70 @@ const handleResponseToReview = async (context) => {
+      if (reviewersToPing.length !== 0) {
+        const message = 'Hi' + reviewersToPing.map((reviewer) => {
+          return ' @' + reviewer.login + ' ';
+        }) + ', Its been more than 24 hours, ' +
+         'since the review was requested. Please review this. Thanks!';
```
We are not taking the time when the review was requested but when the review was submitted right?

Naman-1234 replied:
Sorry for that, Will change it in the latest commit Sir. Thanks.

vojtechjelinek replied:
No, this is not correct, What I meant is that the comment was correct, but the code is wrong.

Naman-1234 replied:
Sir, Please could you be more specific about where the code is wrong and how it could be corrected. So I can start working on this. Thanks.

vojtechjelinek replied:
So, in your code you look at all the reviews and if the review is older than 24 hours you notify the reviewer, but that doesn't make sense. We want to notify the reviewer if the REQUEST for review is older than 24 hours.

Naman-1234 replied:
It does make sense, Thanks for pointing it out. Will perform the changes in the latest commits Sir.

Naman-1234 replied:
@vojtechjelinek  and @jameesjohn I am thinking of the following approach, Could you please confirm whether this is alright.
- Check in the list of requested reviewers.
- Then mapping on to that list and noticing their last reviews. If their review was done before 24 HRS, then ping them .

Does this approach look alright to you sir. Please let me know so i can proceed in this PR.Thanks.

vojtechjelinek replied:
That is still the same solution right? The solution should be 'if the review was requested more than 24 hours ago and the reviewer didn't do anything on the PR notify them'.

Naman-1234 replied:
@jameesjohn Sir Could you help me out with how can I find the time for the requested review, I didn't find anything related to it. It would be a big help if you could guide me on this one. Thanks.

jameesjohn replied:
@vojtechjelinek, we may not be able to get exactly when the review got requested from the pull request data. 
But we can assume that all the currently assigned need to review the pull request, and we can check if they have performed the review or check if they have commented on the pull request. 
Then if they haven't performed any of those, we notify them.

What do you think?

Naman-1234 replied:
@jameesjohn Sir According to me, this approach is the same as currently being implemented by me. Am I right?. Thanks.

jameesjohn replied:
Yes, I think so.

Naman-1234 replied:
@vojtechjelinek, Sir Could you please reply to the same conversation. Thanks.

vojtechjelinek reviewed (changes requested):
Thanks. Two comments.

Naman-1234 commented:
@vojtechjelinek Sir Could you please confirm if the changes are fine now. Thank you.

vojtechjelinek commented:
@Naman-1234 @jameesjohn I'm still unsure if this is the exact behaviour that we want. Could you please test this PR on some testing repository and PR and then show me the result?

The tests should be:

1. PR gets created
2. Reviewer reviews it
3. 30 hours pass (Oppiabot shouldn't do anything)
4. Author fixes the code and requests a new review from the original reviewer
5. 30 hours pass (Oppiabot pings the reviewer)

Naman-1234 commented:
@vojtechjelinek  Sir Actually the behaviour of the Oppiabot would be as follows according to the condition given by you:

- It will not run on PR Creation as it is a Periodic Check (This could be changed if you want to a PR Created event).
- Reason behind it was made a periodic Check was that till now Github API(On which probot is based) does not provide any method of checking when the review got requested. Also confirmed by James Sir here https://github.com/oppia/oppiabot/pull/215#discussion_r619138659.
-  This is what the current solution gives:
    1.It checks after every 24 HRS(Time of Periodic Check).
    2. It checks all the reviews of the Pull Request and checks the last review of all reviewers. If that review was done for more than 24 Hrs before, then it mentions that reviewer to review this PR. 
-  Does this provide any help on the answer to tests asked by you.  Please let me know.

@jameesjohn Sir Could you help me a little bit here how can I modify this PR to bring to the solution Sir is suggesting or is this the best solution which could be implemented now. Could you please confirm on this one. Looking forward for your reply. Thanks.

jameesjohn commented:
> 4. Author fixes the code and requests a new review from the original reviewer
> 5. 30 hours pass (Oppiabot pings the reviewer)

@Naman-1234 @vojtechjelinek  To achieve this, we need to have some sort of database solution which we use to track the time the review was requested.

vojtechjelinek commented:
Could you use GitHub API (https://api.github.com/users/vojtechjelinek/events) for retrieving the events that the user has done? Basically pinging a user that already revied something and is only waiting for the author to address the comments is useless. We somehow need to get at least an approximate time when the review was requested and only notify reviewers that the author is waiting on.

Naman-1234 commented:
@vojtechjelinek  Looking into that Sir. Thanks.

Naman-1234 commented:
@vojtechjelinek Sir I looked into that, But in that also there is no event of pull Request review but there were events of Push, pullRequestComment, I was thinking of a new approach, Could you please confirm if that looks nice. If it does I can start working on now.

New Approach:
- Check for Push event.
- Check for commentEvent after push.
- If comment contains the term PTAL or review etc. Then ping code-owners after 24 hrs of that comment time.

@jameesjohn  and @vojtechjelinek  Could you please confirm about the approach. Thanks.
